docs/references/architecture/adr-113-mempool-lanes.md

```diff
@@ -0,0 +1,312 @@
+# ADR 113: Mempool Lanes
+
+## Changelog
+
+- 2024-04-12: Initial notes (@hvanz)
+- 2024-04-12: Comments on the notes (@hvanz)
+- 2024-04-17: Discussions (@sergio-mena @hvanz)
+- 2024-04-18: Preliminary structure (@hvanz)
+
+## Status
+
+In writing (we're not sure yet whether this will be an ADR or a spec)
+
+## Context
+
+TODO
+
+Try using "QoS" term. Because that's what it actually is.
+
+We need problem definition.
+
+
+Definitions
+Tx1, Tx2
+Traffic classes
+Traffic class order (partial?)
+Latency(tx) --- important
+
+Properties / Requirements (similar to abci++ requirement)
+
+[Examples]
+* Property 1: tx1 --> tc1, tx2 --> tc1 ---> FIFO
+
+* Property 2: tx1 --> tc1, tx2 --> tc2; tc1 < tc2 --> tx1 first, then tx2
+
+* Property 3: tx1 --> tc1, tx2 --> tc2; tc < tc2 --> latency(tx1)<=latency(tx2)
+
+Try to come up with properties:
+* Currently, these properties not guaranteed
+* The rest of this ADR, proposal to guarantee these properties
+
+Note: Consider explaining that the problem we are trying to solve is **not** general bandwidth or latency improvements: that is an orthogonal problem.
+
+## Alternative Approaches
+
+TODO
+
+Prio mempool: why is it not a good fit?
+
+Other alternatives:
+* Consider looking into mempool discussions in Solana
+* Ethereum pre-confirmations
+* Timebox this: 1-2 days each
+
+
+## Decision
+
+TODO
+
+## Detailed Design
+
+TO DECIDE:
+
+* bytes per tx, \# of txs, total # of bytes: shared?
+  * Current view: **can be different** easily
+  * MVP is viable without it, but use cases less customizable
+  * Update: How about Injective's use case?
+
+* Two options for introducing several TX lists:
+  1. Multiple `Clist`s inside `CListMempool`, the cache stays inside `CListMempool`.
+  2. Multiple `CListMempool` instances, the cache needs to be outside `CListMempool` (likely, directly controlled by the reactor)
+
+* After discussion with Osmosis and Injective, there's a dilemma.
+  * For Osmosis, a two-lane solution ("native" and "priority") would probably be enough.
+    * This means that we could defer the lane and priority configuration, and not tackle it as part of the MVP
+  * For Injective, their use case seems to be more demanding: many lanes (4?, 9?),
+    with one lane even having less prio than the native lane
+    * If we want to address that use case fully, our MVP _does need_ to tackle the design of lane & prio configuration
+    * so, a more ambitious MVP, implying more energy, time, and resources. We must be careful with what we promise for when
+
+* Byzantine nodes may gossip transactions on a high-priority lane to get preferential treatment.
+  * we may consider banning peers that send transactions with incorrect lane information.
+  * Decision was not to ship lane info, but we're still using different channels for different lanes
```

Comment:
So we have a variable number of channels? Notice that channel IDs must match in all instances, and they can't collide.

Reply:
Yes, we have a solution to that below, which we will have to present to users to make sure it addresses their use cases